sys/mips/rootdev.c

```diff
@@ -2,73 +2,168 @@
+  rman_t mem, irq;
+  intr_event_t intr_event[8];
 } rootdev_t;
 
-static rman_t rm_mem; /* stores all resources of root bus children */
+static void rootdev_mask_irq(intr_event_t *ie) {
+  int irq = ie->ie_irq;
+  mips32_bc_c0(C0_STATUS, SR_IM0 << irq);
+}
+
+static void rootdev_unmask_irq(intr_event_t *ie) {
+  int irq = ie->ie_irq;
+  mips32_bs_c0(C0_STATUS, SR_IM0 << irq);
+}
 
-static void rootdev_intr_setup(device_t *dev, unsigned num,
+static void rootdev_intr_setup(device_t *dev, unsigned irq,
                                intr_handler_t *handler) {
-  mips_intr_setup(handler, num);
+  rootdev_t *rd = dev->parent->state;
+  intr_event_t *event = &rd->intr_event[irq];
+  intr_event_add_handler(event, handler);
 }
 
 static void rootdev_intr_teardown(device_t *dev, intr_handler_t *handler) {
-  mips_intr_teardown(handler);
+  intr_event_remove_handler(handler);
 }
 
-static int rootdev_attach(device_t *bus) {
-  /* Manages space occupied by I/O devices: PCI, FPGA, system controler, ... */
-  rman_init(&rm_mem, "Malta I/O space", 0x10000000, 0x1fffffff, RT_MEMORY);
-  device_add_child(bus, NULL, 0); /* for GT PCI */
-  return bus_generic_probe(bus);
-}
+static resource_t *rootdev_alloc_resource(device_t *dev, res_type_t type,
+                                          int rid, rman_addr_t start,
+                                          rman_addr_t end, size_t size,
+                                          res_flags_t flags) {
+  rootdev_t *rd = dev->parent->state;
+  rman_t *rman = NULL;
+  resource_t *r;
 
-static resource_t *rootdev_alloc_resource(device_t *bus, device_t *child,
-                                          res_type_t type, int rid,
-                                          rman_addr_t start, rman_addr_t end,
-                                          size_t size, res_flags_t flags) {
+  if (type == RT_MEMORY)
+    rman = &rd->mem;
+  else if (type == RT_IRQ)
+    rman = &rd->irq;
+  else
+    panic("Resource type not handled!");
 
-  resource_t *r =
-    rman_alloc_resource(&rm_mem, start, end, size, 1, RF_NONE, child);
+  if (!(r = rman_alloc_resource(rman, start, end, size, 1, RF_NONE)))
+    return NULL;
 
-  if (r) {
+  if (type == RT_MEMORY) {
     r->r_bus_tag = generic_bus_space;
-    if (flags & RF_ACTIVE)
-      bus_space_map(r->r_bus_tag, r->r_start, r->r_end - r->r_start + 1,
-                    &r->r_bus_handle);
+    r->r_bus_handle = r->r_start;
+
+    if (flags & RF_ACTIVE) {
+      if (bus_activate_resource(dev, type, rid, r)) {
+        rman_release_resource(r);
+        return NULL;
+      }
+    }
   }
 
   return r;
 }
 
+static void rootdev_release_resource(device_t *dev, res_type_t type, int rid,
+                                     resource_t *r) {
+  panic("not implemented!");
+}
+
+static int rootdev_activate_resource(device_t *dev, res_type_t type, int rid,
+                                     resource_t *r) {
+  int error = 0;
+
+  if (r->r_flags & RF_ACTIVE)
+    return 0;
```

Comment:
Maybe we should return something different than 0, to indicate that we tried to activate a resource for the second time? I believe code that ignores the fact that it tried to activate an already activated resource might break something.

Reply:
Can you give a specific scenario that something will break? If a resource is activated, then activating it the second time is no-op, isn't it?